File: cc_Wallet/Wallet.js
const promptSync = require('./PromptSync');
const walletFunc = require('./WalletFunc');

class Wallet {
  constructor() {
    this.items = [];
  }

  displayMenu(title, options) {
    const MAIN_MENU_OPTIONS = ['Manage PIT', 'Manage Money', 'Display All Items', 'Exit'];
    const PIT_MENU_OPTIONS = ['Add NIC', 'Add Banking Card', 'Add Driving Licence', 'Add Visit Card', 'Add Identification Photo', 'Back to Main Menu'];
    const WALLET_MENU_OPTIONS = ['Add Money', 'View Balance', 'Withdraw Money', 'Back to Main Menu'];

    console.log(`\n${title}:`);
    options.forEach((option, index) => console.log(`${index + 1}. ${option}`));

    return { MAIN_MENU_OPTIONS, PIT_MENU_OPTIONS, WALLET_MENU_OPTIONS };
  }

  async mainMenu() {
    let continueFlag = true;

    while (continueFlag) {
      const { MAIN_MENU_OPTIONS } = this.displayMenu('Main Menu', ['Manage PIT', 'Manage Money', 'Display All Items', 'Exit']);

      const option = await promptSync('Choose an option (1/2/3/4): ');

      switch (option) {
        case '1':
          await this.manageOptions('Manage PIT Menu', ['Add NIC', 'Add Banking Card', 'Add Driving Licence', 'Add Visit Card', 'Add Identification Photo', 'Back to Main Menu'], this.addPIT.bind(this));
          break;
        case '2':
          await this.manageOptions('Manage Wallet', ['Add Money', 'View Balance', 'Withdraw Money', 'Back to Main Menu'], this.executeWalletOption.bind(this));
          break;
        case '3':
          this.displayAllItems();
          break;
        case '4':
          this.exit();
          break;
        default:
          console.log('Invalid option. Please choose 1, 2, 3, or 4.');
      }
    }
  }

  async manageOptions(menuTitle, options, action) {
    const { MAIN_MENU_OPTIONS } = this.displayMenu(menuTitle, options);

    const option = await promptSync(`Choose an option (1/${options.length}): `);

    if (option >= 1 && option <= options.length) {
      await action(options[option - 1]);
    } else if (option === options.length + 1) {
      console.log('Returning to Main Menu.');
    } else {
      console.log(`Invalid option. Please choose a number between 1 and ${options.length + 1}.`);
    }
  }
  async addPIT(pitType) {
    const result = walletFunc.addPITToWallet(this.items, pitType);
    result.messages.forEach(message => console.log(message));
  }

  displayAllItems() {
    console.log('\nAll Items in the Wallet:');

    let totalBalance = 0;
    let lastAction = 'added';

    this.items.forEach((item, index) => {
      console.log(`-------------------${item.type === 'pit' ? 'PIT' : 'MONEY'}-----------------------`);
      console.log(`${item.type === 'pit' ? 'PIT' : 'Money'} ${index + 1}: ${item.type === 'pit' ? item.pitType : `Ar${item.amount.toFixed(2)} ${item.amount > 0 ? '"added"' : '"withdrawn"'}`}`);
      
      if (item.type === 'money') {
        totalBalance += item.amount;
      }
    });

    console.log(`\nTotal Balance: Ar${totalBalance.toFixed(2)} "${lastAction}"`);
  }

  async executeWalletOption(option) {
    switch (option) {
      case 'Add Money':
        await this.addMoney();
        break;
      case 'View Balance':
        await this.viewMoney();
        break;
      case 'Withdraw Money':
        await this.withdrawMoney();
        break;
      case 'Back to Main Menu':
        console.log("------*---------*------");
        console.log('Returning to Main Menu.');
        console.log("------*---------*------");
        break;
      default:
        console.log('Invalid Wallet option. Please choose Add Money, View Balance, Withdraw Money, or Back to Main Menu.');
    }
  }

  async addMoney() {
    const amountInput = await promptSync('Enter the amount to add: ');
    const amount = parseFloat(amountInput);
    const result = walletFunc.addMoneyToWallet(this.items, amount);
    result.messages.forEach(message => console.log(message));
  }

  async viewMoney() {
    const result = walletFunc.viewMoneyInWallet(this.items);
    result.messages.forEach(message => console.log(message));
  }

  async withdrawMoney() {
    const amountInput = await promptSync('Enter the amount to withdraw: ');
    const amount = parseFloat(amountInput);
    const result = walletFunc.withdrawMoneyFromWallet(this.items, amount);
    result.messages.forEach(message => console.log(message));
  }

  exit() {
    console.log('Exiting the application. Goodbye!');
    process.exit();
  }
}

module.exports = Wallet;

File: cc_Wallet/wallet.py
from wallet_func import WalletFunc
import asyncio

class Wallet:
    def __init__(self):
        self.items = []

    def display_menu(self, title, options):
        print(f'\n{title}:')
        for index, option in enumerate(options, start=1):
            print(f'{index}. {option}')

    async def main_menu(self):
        continue_flag = True
        while continue_flag:
            self.display_menu('Main Menu', ['Manage PIT', 'Manage Money', 'Display All Items', 'Exit'])
            option = input('Choose an option (1/2/3/4): ')
            if option == '1':
                await self.manage_pit()
            elif option == '2':
                await self.manage_money()
            elif option == '3':
                self.display_all_items()
            elif option == '4':
                self.exit()
            else:
                print('Invalid option. Please choose 1, 2, 3, or 4.')

    async def manage_options(self, menu_title, options, action):
        self.display_menu(menu_title, options)
        option = input(f'Choose an option (1/{len(options)}): ')
        if option.isdigit() and 1 <= int(option) <= len(options):
            await action(options[int(option) - 1])
        elif option == str(len(options) + 1):
            print('Returning to Main Menu.')
        else:
            print(f'Invalid option. Please choose a number between 1 and {len(options) + 1}.')

    async def manage_pit(self):
        pit_options = ['Add NIC', 'Add Banking Card', 'Add Driving Licence', 'Add Visit Card', 'Add Identification Photo', 'Back to Main Menu']
        await self.manage_options('Manage PIT Menu', pit_options, self.add_pit)

    async def add_pit(self, pit_type):
        result = WalletFunc().add_pit_to_wallet(self.items, pit_type)
        for message in result['messages']:
            print(message)

    def display_all_items(self):
        print('\nAll Items in the Wallet:')
        total_balance = 0
        last_action = 'added'
        for index, item in enumerate(self.items, start=1):
            print(f'-------------------{"PIT" if item["type"] == "pit" else "MONEY"}-----------------------')
            if item['type'] == 'pit':
                print(f'PIT {index}: {item["pit_type"]}')
            elif item['type'] == 'money':
                action = 'added' if item['amount'] > 0 else 'withdrawn'
                print(f'Money {index}: Ar{item["amount"]:.2f} {action}')
                total_balance += item['amount']
        print(f'\nTotal Balance: Ar{total_balance:.2f} {last_action}')

    async def manage_money(self):
        wallet_options = ['Add Money', 'View Balance', 'Withdraw Money', 'Back to Main Menu']
        await self.manage_options('Manage Wallet', wallet_options, self.execute_wallet_option)

    async def execute_wallet_option(self, option):
        if option == 'Add Money':
            await self.add_money()
        elif option == 'View Balance':
            self.view_money()
        elif option == 'Withdraw Money':
            await self.withdraw_money()
        elif option == 'Back to Main Menu':
            print("------*---------*------")
            print('Returning to Main Menu.')
            print("------*---------*------")
        else:
            print('Invalid Wallet option. Please choose Add Money, View Balance, Withdraw Money, or Back to Main Menu.')

    async def add_money(self):
        amount = float(input('Enter the amount to add: '))
        result = WalletFunc().add_money_to_wallet(self.items, amount)
        for message in result['messages']:
            print(message)
        self.items.append({'type': 'money', 'amount': amount})

    def view_money(self):
        result = WalletFunc().view_money_in_wallet(self.items)
        for message in result['messages']:
            print(message)

    async def withdraw_money(self):
        amount = float(input('Enter the amount to withdraw: '))
        result = WalletFunc().withdraw_money_from_wallet(self.items, amount)
        for message in result['messages']:
            print(message)
        if result['success']:
            self.items.append({'type': 'money', 'amount': -amount})

    def exit(self):
        print('Exiting the application. Goodbye!')
        exit()

async def run():
    wallet = Wallet()
    while True:
        await wallet.main_menu()

if __name__ == "__main__":
    asyncio.run(run())
